### PR TITLE
#1-4 Task. Punko Pavel

### PR DESCRIPTION
**Pull request by PashaPunko**
https://travis-ci.org/github/PashaPunko/js-assignments/builds/706069585

### CODE REVIEW

File: task/04-arrays-tasks.js [hunk 1]
```diff
@@ -519,7 +541,8 @@ function group(array, keySelector, valueSelector) {
  *   ['one','two','three'], x=>x.split('')  =>   ['o','n','e','t','w','o','t','h','r','e','e']
  */
 function selectMany(arr, childrenSelector) {
-    throw new Error('Not implemented');
+    arr = arr.map(x=>childrenSelector(x));
+    return arr.reduce((acc, val) => acc.concat(val), []);
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: this is alright, but you could use childrenSelector inside reduce and do the same in one pass

File: task/04-arrays-tasks.js [hunk 2]
```diff
@@ -536,7 +559,8 @@ function selectMany(arr, childrenSelector) {
  *   [[[ 1, 2, 3]]], [ 0, 0, 1 ]      => 2        (arr[0][0][1])
  */
 function getElementByIndexes(arr, indexes) {
-    throw new Error('Not implemented');
+    if (indexes.length == 1) return arr[indexes[0]];
+    else return getElementByIndexes(arr[indexes[0]], indexes.slice(1))
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js]: works but it's better to not use recursion if possible. Try to do it with reduce function

File: task/04-arrays-tasks.js [hunk 3]
```diff
@@ -559,7 +583,10 @@ function getElementByIndexes(arr, indexes) {
  * 
  */
 function swapHeadAndTail(arr) {
-    throw new Error('Not implemented');
```
Comment (AlexMokin) [inline task/04-arrays-tasks.js:562]: looks like a "swap two numbers without using a temp" problem. Only you're not saving anything here, each map creates new array.  Try to find simpler solution. Hint: use concat and slice

**Comment by AlexMokin** [comment]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by PashaPunko** [comment]
https://travis-ci.org/github/PashaPunko/js-assignments/builds/706069585

**Comment by AlexMokin** [comment]
Please carefully review the commetns and make the fixes according to the comments.

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by PashaPunko** [comment]
Good evening. There was a wrong link on travis, probably. I fixed it and each tasks are complete. So if you mean this issue in your comments, can I mark this PR as ready for review?

**Comment by AlexMokin** [comment]
I mean i left notes in the comments. The code must be fixed according with the comments. You can find comments on the **conversation** and **Files changed** tabs

E.g.

image

I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by PashaPunko** [comment]
> I mean i left notes in the comments. The code must be fixed according with the comments. You can find comments on the conversation and Files changed tabs
> E.g.
> 
> I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

screencapture-github-AisBrestEDU-js-assignments-pull-333-2020-07-11-15_42_06

That's a problem. As you may see at the screen short, there is no comments you left in conversation.

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -201,7 +201,19 @@ function extractEmails(str) {
...
-    throw new Error('Not implemented');
+    const RT_ANGLE = '\u2510';
+    const RB_ANGLE = '\u2518';
+    const LT_ANGLE = '\u250C';
+    const LB_ANGLE = '\u2514';
+    const H_LINE = '\u2500';
+    const V_LINE = '\u2502';
+    let str = '';
+    if (width > 1 && height>1){
+        str = LT_ANGLE + new String(H_LINE).repeat(width-2) + RT_ANGLE+'\n';
```
Comment (AlexMokin) [inline task/01-strings-tasks.js]: H_LINE is already a string you don't need to wrap it in new String()

**Review by AlexMokin — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by AlexMokin** [comment]
accepted

**Pull request closed without merging**